neicnordic/sensitive-data-archive
# Feat: connect reencrypt service to sda-download

**Related issue(s) and PR(s)**  
This PR closes #698 

**Description**
With this PR, the users can reencrypt the header of the crypt4gh encrypted file by providing the client public key at the request header with the key `client-public-key`. 

**How to test**
1. Build the image for sda-download
In the sub-folder `sda-download`, run 
```
docker build . -t local-sda-download
```
 
2.  Set up the docker environment with [this branch](https://github.com/GenomicDataInfrastructure/starter-kit-storage-and-interfaces/tree/feat/reencrypt-header) of the repo `starter-kit-storage-and-interfaces`
with the command 
```
cp .env.example .env
cp config/config.yaml.example config/config.yaml
docker compose -f docker-compose-full-demo-test.yml  up -d 
```
3.  Create your client key pair by 
```
sda-cli createKey client
```
4. Obtain the token by 
```
token=$(curl -s -k https://localhost:8080/tokens | jq -r '.[0]') 
```
5. Get base64 encoded client key by 
```
clientkey=$(base64 -w0 client.pub.pem) 
```
6. Download the test BAM file with the client public key 
```
curl -v -H "Authorization: Bearer $token" -H "Client-Public-Key: $clientkey" http://localhost:8443/s3-encrypted/$dataset/NA12878.bam -o reenc.bam 
```
7. Verify by decrypt it with your client key and view the content
```
crypt4gh decrypt -s client.sec.pem -f reenc.bam
samtools view reenc.bam.dec
```
**Additional notes**
* The code for the reencrypt service has been copied (and thus duplicated) from `sda/internal/reencrypt` to `sda-download/internal/reencrypt` since it is not exported as public. This needs to be removed once `sda-download` is merged.
